### ISSUE TITLE
Should OmniSharp be in charge of file watching?

### ISSUE THREAD

**Issue by jrieken** [OmniSharp/omnisharp-roslyn#106]
I wonder if OmniSharp should be watching files that contribute to the state of a project. Such files are  ```*.cs```-files but also ```project.json```, ```*.proj``` etc. Basically, everything that is loaded (or required) to build up the project state during ```IProjectSystem#Initialize```. My reasoning for doing this in OmniSharp is the following:
* file watching is hard and if it's in OmniSharp everyone can benefit
* editors usually don't know what files a project is build up from but make qualified guesses (glob-patterns etc)
* it would make routes like ```/reloadSolution``` not needed anymore

Also, there is a differentiation between files that are _open in the editor_ and those that are not. For instance, editor plugins send buffer updates for ```*.cs``` files but not for ```project.json```. So maybe, once a file has been receiving updates via the ```/updatebuffer``` or ```/changebuffer``` route it should not be watched any more. 

Another question to which I don't know an answer: How good is the mono file watcher?

**Comment by nosami** [comment]
I think in a perfect world, we would handle this on the server. 

However, the mono file watcher is currently terrible in 3.12 (but is currently being worked on). @davidfowl had to remove some file watching code because CPU usage frequently spiked to 100%

Also, I don't see why the editor needs to know what project a file belongs to. None of the editors are project aware. 

Although the editors don't update the server with `project.json` today, there is no reason why that couldn't happen.

It's a little frustrating. I agree with you, but the reason I added `/reloadsolution` in the first place was to deal with deficiencies in FSW. It would be really nice to not have to put all this stuff in the editor.

**Comment by davidfowl** [comment]
I added https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp/Services/IFileSystemWatcher.cs a while back so we could get started on something (the interface might need some clean up). 

So lets do this:

- Implement IFileSystemWatcher and all of the logic within each project system (I think some things are there already) and get that working end to end.
- Have the editor watch files in the short term, and send a /fileschanged event for any file add/remove/change (basically a file system watcher replacement). This is ok because the server will filter events based on what it watched.
- When we have a viable replacement, we swap the IFileSystemWatcher implementation.

**Comment by jrieken** [comment]
Great - I will start by adding a ```filesChanged``` route which drives a special file watcher implementation. Next step is to establish a way to communicate back the effects of such changes, like new compile errors when removing a reference etc.

**Comment by jrieken** [comment]
@davidfowl, @nosami - A bit of progress can be found here: https://github.com/OmniSharp/omnisharp-roslyn/tree/filesChanged. Generally, I have exposed ```TriggerChange``` via a controller (maybe more refinement about event types?) and in return I send back events the workspace generates (like project changed, document added etc). The project systems still lack proper handling of file changes, etc when a new ```cs```-file is being added. It also lacks an event emitter for http (long poll?)

**Comment by nosami** [comment]
I wouldn't worry about the HTTP side. Long polling sounds a little yucky and we'll probably start to migrate the editors over to stdio at some point.

I know that not all editors are created equally, but I don't quite see why the editor even needs to know that a project has been added/changed/reloaded. The thing I like most about using OmniSharp is that I don't even have to think about projects any more - everything is just a file on disk or in memory in an editor buffer.

**Comment by jrieken** [comment]
Yeah - don't really mind if we include project events or not. My main concern is events for documents. However, sending events for projects will get you around the pretty broken ```/projects```-route. Currently, that is indeterministic as its result depend on response messages from DTH (for which we don't wait).

**Comment by nosami** [comment]
Yeah... that's a good point. I'd forgotten about that :)
